File: alife/brain.py
# brain.py
# Спайковая нейронная сеть (SNN) агента
import numpy as np
from .config import INPUT_SIZE, OUTPUT_SIZE, LEARNING, SYNAPTIC_SCALE, clamp


class Brain:
    def __init__(self, genome, n_hidden=None, parent_weights=None):
        # Наследуемая архитектура мозга - количество скрытых нейронов из генома
        if n_hidden is None:
            from .config import N_HIDDEN
            self.n_hidden = N_HIDDEN
        else:
            self.n_hidden = n_hidden
        
        self.n_in = INPUT_SIZE
        self.n_out = OUTPUT_SIZE
        self.n = self.n_in + self.n_hidden + self.n_out
        self.hidden_slice = slice(self.n_in, self.n)
        self.v = np.zeros(self.n, dtype=np.float32)
        self.spikes = np.zeros(self.n, dtype=np.float32)
        self.out_rate = np.zeros(self.n_out, dtype=np.float32)

        rng = np.random.default_rng()
        self.mask = rng.random((self.n, self.n)) < genome["conn_prob"]
        np.fill_diagonal(self.mask, False)

        in_mask = rng.random((self.n_in, self.n)) < max(genome["conn_prob"], 0.12)
        self.mask[:self.n_in, :] |= in_mask
        out_mask = rng.random((self.n, self.n_out)) < max(genome["conn_prob"], 0.15)
        self.mask[:, -self.n_out:] |= out_mask
        np.fill_diagonal(self.mask, False)

        self.W = (
            rng.normal(0.0, genome["weight_scale"], (self.n, self.n)).astype(np.float32)
            * self.mask
        )

        if parent_weights is not None and parent_weights.shape == self.W.shape:
            lam = clamp(genome["lamarckian_weight"], 0.0, 1.0)
            self.W = ((1.0 - lam) * self.W + lam * parent_weights).astype(np.float32)
            self.W *= self.mask

        self.decay_base = genome["membrane_decay"]
        self.threshold_base = genome["threshold"]
        self.stdp_rate = genome["stdp_rate"]
        self.max_w = genome["weight_max"]

        if LEARNING:
            self.E = np.zeros((self.n, self.n), dtype=np.float32)
        else:
            self.E = None

    def step(self, sensors, mod):
        sensors = np.asarray(sensors, dtype=np.float32)
        pre = self.spikes
        current = pre @ self.W

        arousal = mod.get("arousal", 0.0)
        decay = clamp(self.decay_base + arousal * 0.02, 0.50, 0.99)
        threshold = clamp(self.threshold_base - arousal * 0.05, 0.30, 2.0)

        self.v = self.v * decay + current * SYNAPTIC_SCALE

        if arousal > 0.8:
            noise_size = self.n - INPUT_SIZE
            noise = np.random.normal(0.0, (arousal - 0.8) * 0.02, noise_size).astype(np.float32)
            self.v[INPUT_SIZE:] += noise

        new_spikes = np.zeros(self.n, dtype=np.float32)
        hidden_v = self.v[self.hidden_slice]
        fired = hidden_v >= threshold
        new_spikes[self.hidden_slice] = fired.astype(np.float32)
        hidden_v[fired] = 0.0

        new_spikes[:INPUT_SIZE] = np.clip(sensors, 0.0, 1.0)
        self.out_rate = 0.75 * self.out_rate + 0.25 * new_spikes[-self.n_out:]

        if LEARNING and self.E is not None:
            learn_rate = clamp(
                mod.get("plasticity", 0.0) * mod.get("dopamine", 0.0),
                -2.0, 2.0,
            )
            if abs(learn_rate) > 1e-6:
                post = new_spikes
                delta = (np.outer(pre, post) - np.outer(post, pre)).astype(np.float32)
                self.E = self.E * 0.95 + delta * self.stdp_rate
                self.W += learn_rate * self.E * self.mask
                self.W = np.clip(self.W, -self.max_w, self.max_w)
                self.W *= self.mask

        self.spikes = new_spikes
        return self.out_rate

File: cpp/src/brain.cpp
// brain.cpp
// Спайковая нейронная сеть (SNN) агента - реализация
#include "alife/brain.h"
#include "alife/math_utils.h"
#include <cmath>
#include <stdexcept>

namespace alife {

Brain::Brain(
    const Genome& genome,
    std::optional<int> n_hidden_opt,
    const std::optional<std::vector<double>>& parent_weights,
    RNG* rng
) : rng_(rng ? rng : &default_rng_),
    n_in(DEFAULT_INPUT_SIZE),
    n_out(DEFAULT_OUTPUT_SIZE),
    n_hidden(n_hidden_opt.value_or(DEFAULT_N_HIDDEN)),
    n(n_in + n_hidden + n_out),
    decay_base(genome.get("membrane_decay", 0.85)),
    threshold_base(genome.get("threshold", 1.0)),
    stdp_rate(genome.get("stdp_rate", 0.01)),
    max_w(genome.get("weight_max", 2.0)),
    learning(true)
{
    // Инициализация массивов
    v.assign(n, 0.0);
    spikes.assign(n, 0.0);
    out_rate.assign(n_out, 0.0);
    
    // Маска связей и веса
    size_t total_size = static_cast<size_t>(n) * static_cast<size_t>(n);
    mask.assign(total_size, false);
    W.assign(total_size, 0.0);
    
    // Генерация маски соединений
    double conn_prob = genome.get("conn_prob", 0.1);
    double weight_scale = genome.get("weight_scale", 0.5);
    
    // Основная маска соединений
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i == j) continue;  // Диагональ всегда False
            
            if (rng_->next_float() < conn_prob) {
                mask[idx(i, j)] = true;
                W[idx(i, j)] = rng_->gauss(0.0, weight_scale);
            }
        }
    }
    
    // Усиленные связи от входов
    double in_conn_prob = std::max(conn_prob, 0.12);
    for (int i = 0; i < n_in; ++i) {
        for (int j = 0; j < n; ++j) {
            if (rng_->next_float() < in_conn_prob) {
                mask[idx(i, j)] = true;
                if (W[idx(i, j)] == 0.0) {
                    W[idx(i, j)] = rng_->gauss(0.0, weight_scale);
                }
            }
        }
    }
    
    // Усиленные связи к выходам
    double out_conn_prob = std::max(conn_prob, 0.15);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n_out; ++j) {
            int col = n - n_out + j;
            if (rng_->next_float() < out_conn_prob) {
                mask[idx(i, col)] = true;
                if (W[idx(i, col)] == 0.0) {
                    W[idx(i, col)] = rng_->gauss(0.0, weight_scale);
                }
            }
        }
    }
    
    // Снова обнуляем диагональ
    for (int i = 0; i < n; ++i) {
        mask[idx(i, i)] = false;
        W[idx(i, i)] = 0.0;
    }
    
    // Наследование родительских весов
    if (parent_weights.has_value() && 
        parent_weights->size() == total_size) {
        double lamarckian = clamp(
            genome.get("lamarckian_weight", 0.0),
            0.0, 1.0
        );
        
        for (size_t i = 0; i < total_size; ++i) {
            W[i] = ((1.0 - lamarckian) * W[i] + lamarckian * (*parent_weights)[i]);
            if (!mask[i]) {
                W[i] = 0.0;
            }
        }
    }
    
    // Eligibility trace только если обучение включено
    int total_neurons = n_in + n_hidden + n_out;
    if (learning && total_neurons <= 1200) {
        E.assign(total_size, 0.0);
    } else {
        learning = false;
        E.clear();
    }
}

std::vector<double> Brain::step(
    const std::vector<double>& sensors,
    const std::map<std::string, double>& mod
) {
    // Проверка размера сенсоров
    if (sensors.size() != static_cast<size_t>(n_in)) {
        throw std::invalid_argument(
            "Ожидается " + std::to_string(n_in) + " сенсоров, получено " + 
            std::to_string(sensors.size())
        );
    }
    
    // pre = spikes (предыдущие спайки)
    std::vector<double> pre = spikes;
    
    // current = pre @ W (матричное умножение)
    std::vector<double> current(n, 0.0);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            current[i] += pre[j] * W[idx(j, i)];
        }
    }
    
    // Получение модуляторов
    double arousal = 0.0;
    auto it_arousal = mod.find("arousal");
    if (it_arousal != mod.end()) {
        arousal = it_arousal->second;
    }
    
    double plasticity = 0.0;
    auto it_plasticity = mod.find("plasticity");
    if (it_plasticity != mod.end()) {
        plasticity = it_plasticity->second;
    }
    
    double dopamine = 0.0;
    auto it_dopamine = mod.find("dopamine");
    if (it_dopamine != mod.end()) {
        dopamine = it_dopamine->second;
    }
    
    // Вычисление затухания и порога с учётом arousal
    double decay = clamp(decay_base + arousal * 0.02, 0.50, 0.99);
    double threshold = clamp(threshold_base - arousal * 0.05, 0.30, 2.0);
    
    // Обновление мембранного потенциала
    for (int i = 0; i < n; ++i) {
        v[i] = v[i] * decay + current[i] * DEFAULT_SYNAPTIC_SCALE;
    }
    
    // Добавление шума при высоком arousal
    if (arousal > 0.8) {
        double noise_std = (arousal - 0.8) * 0.02;
        for (int i = n_in; i < n; ++i) {
            v[i] += rng_->gauss(0.0, noise_std);
        }
    }
    
    // Генерация новых спайков
    std::vector<double> new_spikes(n, 0.0);
    
    // Скрытые нейроны: проверка порога
    int hidden_start = n_in;
    int hidden_end = n_in + n_hidden;
    for (int i = hidden_start; i < hidden_end; ++i) {
        if (v[i] >= threshold) {
            new_spikes[i] = 1.0;
            v[i] = 0.0;  // Сброс после спайка
        }
    }
    
    // Входные нейроны: просто копируем сенсоры
    for (int i = 0; i < n_in; ++i) {
        new_spikes[i] = clamp(sensors[i], 0.0, 1.0);
    }
    
    // Выходные rates: экспоненциальное скользящее среднее
    for (int i = 0; i < n_out; ++i) {
        int spike_idx = n - n_out + i;
        out_rate[i] = 0.75 * out_rate[i] + 0.25 * new_spikes[spike_idx];
    }
    
    // STDP обучение
    if (learning && !E.empty()) {
        double learn_rate = clamp(plasticity * dopamine, -2.0, 2.0);
        
        if (std::abs(learn_rate) > 1e-6) {
            std::vector<double> post = new_spikes;
            
            // Вычисление delta = outer(pre, post) - outer(post, pre)
            // и обновление eligibility trace
            for (int i = 0; i < n; ++i) {
                for (int j = 0; j < n; ++j) {
                    double delta = pre[i] * post[j] - post[i] * pre[j];
                    E[idx(i, j)] = E[idx(i, j)] * 0.95 + delta * stdp_rate;
                }
            }
            
            // Обновление весов
            for (size_t i = 0; i < W.size(); ++i) {
                W[i] += learn_rate * E[i] * (mask[i] ? 1.0 : 0.0);
                W[i] = clamp(W[i], -max_w, max_w);
                if (!mask[i]) {
                    W[i] = 0.0;
                }
            }
        }
    }
    
    // Сохранение новых спайков
    spikes = new_spikes;
    
    return out_rate;
}

} // namespace alife
